server/src/main/java/io/spine/server/event/funnel/AbstractFunnel.java

```diff
@@ -0,0 +1,166 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.event.funnel;
+
+import com.google.protobuf.Any;
+import com.google.protobuf.Message;
+import com.google.protobuf.Timestamp;
+import io.grpc.stub.StreamObserver;
+import io.spine.base.EventMessage;
+import io.spine.base.Time;
+import io.spine.core.Ack;
+import io.spine.core.ActorContext;
+import io.spine.core.Event;
+import io.spine.core.EventContext;
+import io.spine.core.EventId;
+import io.spine.core.TenantId;
+import io.spine.core.UserId;
+import io.spine.protobuf.AnyPacker;
+import io.spine.server.bus.Bus;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.base.Identifier.newUuid;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * The implementation base for {@link EventFunnel}s.
+ *
+ * <p>Posts the given events into a {@link Bus}.
+ *
+ * @param <M>
+ *         type of accepted by the target bus
+ */
+abstract class AbstractFunnel<M extends Message> implements EventFunnel {
+
+    private final Bus<M, ?, ?, ?> bus;
+    private final UserId actor;
+    private final Any producerId;
+    private final StreamObserver<Ack> resultObserver;
+    private final boolean fromExternalSource;
+    private final @Nullable TenantId tenantId;
+
+    AbstractFunnel(@Nullable TenantId tenantId,
+                   Bus<M, ?, ?, ?> bus,
+                   StreamObserver<Ack> resultObserver,
+                   UserId actor,
+                   boolean fromExternalSource) {
+        this.tenantId = tenantId;
+        this.bus = checkNotNull(bus);
+        this.actor = checkNotNull(actor);
+        this.resultObserver = checkNotNull(resultObserver);
+        this.producerId = AnyPacker.pack(actor);
+        this.fromExternalSource = fromExternalSource;
+    }
+
+    /**
+     * Transforms the given event into the type accepted by the target bus.
+     *
+     * @param event
+     *         the event to post
+     * @return the wrapped event
+     */
+    abstract M transformEvent(Event event);
+
+    /**
+     * Obtains a copy of this funnel but with the given {@code Ack} observer.
+     *
+     * @return new instance of the funnel
+     */
+    abstract AbstractFunnel<M> copyWithObserver(StreamObserver<Ack> resultObserver);
+
+    @Override
+    public final EventFunnel with(StreamObserver<Ack> resultObserver) {
+        checkNotNull(resultObserver);
+        return resultObserver.equals(this.resultObserver)
```

Comment:
Why do we have this optimization here, but not for `forTenant(TenantId)` method?